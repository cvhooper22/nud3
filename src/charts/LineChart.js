import React, { Component, PropTypes } from 'react';
import * as d3 from 'd3';
import _ from 'lodash';

const stringOrArrayOfStrings = PropTypes.oneOfType([
  PropTypes.string,
  React.PropTypes.arrayOf(PropTypes.string),
]);

export default class LineChart extends Component {

  static propTypes = {
    chartData: PropTypes.array,
    className: PropTypes.string,
    classNamePrefix: PropTypes.string,
    colorPalette: PropTypes.any,
    filter: stringOrArrayOfStrings,
    paddingBottom: PropTypes.number,
    paddingLeft: PropTypes.number,
    paddingRight: PropTypes.number,
    paddingTop: PropTypes.number,
    valueKeys: PropTypes.array,
    xScale: PropTypes.func,
    yScale: PropTypes.func,
    clipPath: PropTypes.string,
  };

  constructor (...args) {
    super(...args);
    this.getUniqueDataKey = ::this.getUniqueDataKey;
    this.getStrokeColor = ::this.getStrokeColor;
    this.line = d3.line()
      .x(d => this.props.xScale(d.xValue))
      .y(d => this.props.yScale(d.yValue || 0));
  }

  componentDidMount () {
    this.renderLine();
  }

  componentDidUpdate () {
    this.renderLine();
  }

  getStrokeColor (d, i) {
    if (this.props.colorPalette) {
      if (_.isFunction(this.props.colorPalette)) {
        return this.props.colorPalette(i);
      } else {
        return this.props.colorPalette[i];
      }
    }
    return '';
  }

  getUniqueDataKey (dataSet, i) {
    return this.props.valueKeys[i];
  }

  getPathFilter (d, i) {
    let filter = this.props.filter;
    if (_.isArray(filter)) {
      filter = filter[i];
    }
    return `url(#${filter})`;
  }

  render () {
    const className = [
      'line-chart',
    ];
    if (this.props.classNamePrefix) {
      className.push(`${this.props.classNamePrefix}__line-chart`);
    }
    if (this.props.className) {
      className.push(this.props.className);
    }
    return (
      <g
        className={ className.join(' ') }
        ref={ n => this.node = n }
        transform={ `translate(${this.props.paddingLeft},${this.props.paddingTop})` }
        clipPath={ this.props.clipPath }
      />
    );
  }

  renderLine () {
    this.group = d3.select(this.node);
    const paths = this.group.selectAll('.line-chart__line')
      .data(this.props.chartData, this.getUniqueDataKey)
      .enter()
      .append('path');
    paths
      .attr('class', 'line-chart__line')
      .style('stroke', this.getStrokeColor);
    if (this.props.filter) {
      paths.style('filter', ::this.getPathFilter);
    }


    this.group
      .selectAll('.line-chart__line')
        .attr('d', this.line);
  }
}
