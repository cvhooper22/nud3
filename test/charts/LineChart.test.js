import React from 'react';
import renderer from 'react-test-renderer';
import LineChart from '../../src/charts/LineChart';
import {
  chartData,
  yScale,
  xScale,
  valueKeys,
  originalData,
} from '../mockData/chartData';

describe('LineChart.js', () => {
  it('should take stubbed data and do basic rendering', () => {
    const tree = renderer.create(
      <LineChart
        originalData={ originalData }
        chartData={ chartData }
        areaHeight={ 100 }
        areaWidth={ 100 }
        paddingTop={ 0 }
        paddingBottom={ 0 }
        paddingRight={ 0 }
        paddingLeft={ 0 }
        valueKeys={ valueKeys }
        yScale={ yScale }
        xScale={ xScale }
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('should take stubbed data repositions based on padding', () => {
    const tree = renderer.create(
      <LineChart
        originalData={ originalData }
        chartData={ chartData }
        areaHeight={ 100 }
        areaWidth={ 100 }
        paddingTop={ 10 }
        paddingBottom={ 10 }
        paddingRight={ 10 }
        paddingLeft={ 10 }
        valueKeys={ valueKeys }
        yScale={ yScale }
        xScale={ xScale }
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('should render the clip path', () => {
    const tree = renderer.create(
      <LineChart
        originalData={ originalData }
        chartData={ chartData }
        areaHeight={ 100 }
        areaWidth={ 100 }
        paddingTop={ 0 }
        paddingBottom={ 0 }
        paddingRight={ 0 }
        paddingLeft={ 0 }
        valueKeys={ valueKeys }
        yScale={ yScale }
        xScale={ xScale }
        clipPath="my-clip-path"
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
