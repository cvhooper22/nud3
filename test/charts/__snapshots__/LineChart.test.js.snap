exports[`LineChart.js should render the clip path 1`] = `
<g
  className="line-chart"
  clipPath="my-clip-path"
  transform="translate(0,0)" />
`;

exports[`LineChart.js should take stubbed data and do basic rendering 1`] = `
<g
  className="line-chart"
  clipPath={undefined}
  transform="translate(0,0)" />
`;

exports[`LineChart.js should take stubbed data repositions based on padding 1`] = `
<g
  className="line-chart"
  clipPath={undefined}
  transform="translate(10,10)" />
`;
